examples/ecs/immutable_components.rs

```diff
@@ -0,0 +1,129 @@
+//! This example demonstrates immutable components.
+
+use bevy::{
+    ecs::{component::ComponentId, world::DeferredWorld},
+    prelude::*,
+    utils::HashMap,
+};
+
+/// This component is mutable, the default case. This is indicated by components
+/// implementing two traits, [`Component`], and [`Component<Mutability = Mutable>`].
+#[derive(Component)]
+pub struct MyMutableComponent(bool);
+
+/// This component is immutable. Once inserted into the ECS, it can only be viewed,
+/// or removed. Replacement is also permitted, as this is equivalent to removal
+/// and insertion.
+///
+/// Adding the `#[component(immutable)]` attribute prevents the implementation of [`Component<Mutability = Mutable>`]
+/// in the derive macro.
+#[derive(Component)]
+#[component(immutable)]
+pub struct MyImmutableComponent(bool);
+
+fn demo_1(world: &mut World) {
+    // Immutable components can be inserted just like mutable components.
+    let mut entity = world.spawn((MyMutableComponent(false), MyImmutableComponent(false)));
+
+    // But where mutable components can be mutated...
+    let mut my_mutable_component = entity.get_mut::<MyMutableComponent>().unwrap();
+    my_mutable_component.0 = true;
+
+    // ...immutable ones cannot. The below fails to compile as `MyImmutableComponent`
+    // is declared as immutable.
+    // let mut my_immutable_component = entity.get_mut::<MyImmutableComponent>().unwrap();
+
+    // Instead, you could take or replace the immutable component to update its value.
+    let mut my_immutable_component = entity.take::<MyImmutableComponent>().unwrap();
+    my_immutable_component.0 = true;
+    entity.insert(my_immutable_component);
+}
+
+/// This is an example of a component like [`Name`](bevy::prelude::Name), but immutable.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Component, Reflect)]
+#[reflect(Hash, Component)]
+#[component(
+    immutable,
+    // Since this component is immutable, we can fully capture all mutations through
+    // these component hooks. This allows for keeping other parts of the ECS synced
+    // to a component's value at all times.
+    on_insert = on_insert_name,
+    on_replace = on_replace_name,
+)]
+pub struct Name(pub &'static str);
+
+/// This index allows for O(1) lookups of an [`Entity`] by its [`Name`].
+#[derive(Resource, Default)]
+struct NameIndex {
+    name_to_entity: HashMap<Name, Entity>,
+}
+
+impl NameIndex {
+    fn get_entity(&self, name: &'static str) -> Option<Entity> {
+        self.name_to_entity.get(&Name(name)).copied()
+    }
+}
+
+/// When a [`Name`] is inserted, we will add it to our [`NameIndex`].
+///
+/// Since all mutations to [`Name`] are captured by hooks, we know it is not currently
+/// inserted in the index, and its value will not change without triggering a hook.
+fn on_insert_name(mut world: DeferredWorld<'_>, entity: Entity, _component: ComponentId) {
+    let Some(&name) = world.entity(entity).get::<Name>() else {
+        unreachable!()
```

Comment:
This is more of a Rust question, but why would you prefer this version instead of `unwrap()`? Both need a runtime check (as you aren't using `unreachable_unchecked!()`), both end up being cold code, both will panic with a static message. I'm interested in learning.

Reply:
I think this can be a bit clearer at times, to explain *why* this should never panic. Especially if you add a message in unreachable.

Reply:
Yeah it might be worth adding a message here. The comment explains why but it would probably be clearer to also include a short error message in the `unreachable!`

Reply:
But in this case, wouldn't `.expect(...)` be clearer? I think `unreachable!` should usually be reserved for cases where it isn't immediately obvious from the control flow that a certain code path is unreachable. Unwrapping an `Option` is not one of those cases...

Reply:
I've updated the usage of `unreachable!` in this example to explain why the code is unreachable.

As to `unreachable` vs `expect` vs `panic`, I personally have this hierarchy I like to use:

1. If it should _never_ happen, `unreachable`
2. If I don't _think_ it should happen, `expect`
3. If it _can_ happen but I have no alternative, `panic`

Even though they all produce the same code and behaviour, I view `unreachable` as a stronger assertion than `expect`, and likewise for `panic`. It might be more verbose, but I think it more clearly conveys to anyone reading the code "_Hey I thought about this, it's a bug if this ever panics here_".

But again, just a personal preference, I wouldn't go around changing code to one or the other.

Reply:
This was my reading as well. If my code hits a `panic` I assume my code has a bug. If it hits an `expect` I consider that the library may have a bug. If it hits `unreachable` I begin to wonder if the rust compiler has a bug.

Reply:
Okay, so I haven't missed anything big. Personally, I will not use three lines if one line can do the job.

Reply:
Perfectly reasonable @nakedible! This is just a personal preference of mine.